server/src/main/java/org/elasticsearch/common/xcontent/support/XContentMapValues.java

```diff
@@ -191,62 +192,148 @@ public static Object extractValue(String path, Map<?, ?> map, Object nullValue)
+                // If it's a list, unpack the first list layer to avoid undesired list-of-list representations
+                extractedValues.addAll(replaceNullValues(valueList, nullValue));
+            } else {
+                extractedValues.add(currentValue != null ? currentValue : nullValue);
+            }
+        } else if (currentValue instanceof List<?> valueList) {
+            for (Object o : valueList) {
+                extractAndInsertValue(pathElements, index, o, nullValue, insertNewValue, newValue, extractedValues);
+            }
+        } else if (currentValue instanceof Map<?, ?> map) {
             String key = pathElements[index];
             int nextIndex = index + 1;
-            List<Object> extractedValues = new ArrayList<>();
             while (true) {
                 if (map.containsKey(key)) {
-                    Object mapValue = map.get(key);
-                    if (mapValue == null) {
-                        extractedValues.add(nullValue);
-                    } else {
-                        Object val = extractValue(pathElements, nextIndex, mapValue, nullValue);
-                        if (val != null) {
-                            extractedValues.add(val);
-                        }
+                    // We could use an enum for this, but the scope of use is so limited that an int will do.
+                    // 1 = replace value
+                    // 0 = do nothing
+                    // -1 = remove value
+                    int valueModAction = 0;
+                    if (insertNewValue && nextIndex == pathElements.length) {
+                        // We found a value to extract and then either replace or remove.
+                        // Check the extracted values size before we add to it because if we add an unpacked list, we could potentially
+                        // be adding more than one value.
+                        valueModAction = extractedValues.isEmpty() ? 1 : -1;
+                    }
+
+                    extractAndInsertValue(pathElements, nextIndex, map.get(key), nullValue, insertNewValue, newValue, extractedValues);
+
+                    if (valueModAction == 1) {
+                        // Replace the first instance of the path with the new value
+                        ((Map<String, Object>) map).put(key, newValue);
+                    } else if (valueModAction == -1) {
+                        // Remove any following instances of the path
+                        map.remove(key);
                     }
                 }
                 if (nextIndex == pathElements.length) {
-                    if (extractedValues.size() == 0) {
-                        return null;
-                    } else if (extractedValues.size() == 1) {
-                        return extractedValues.get(0);
-                    } else {
-                        return extractedValues;
-                    }
+                    return;
```

Comment:
What happens if the key was not found? We should still add it and create the missing maps no?

Reply:
Yes, in this case it's added in lines 257-261. We add the value using a flat path to handle if `subojects: false` downstream.